src/lm_polygraph/estimators/focus.py

```diff
@@ -0,0 +1,219 @@
+import os
+import numpy as np
+import torch
+import math
+
+from typing import Dict
+
+from .estimator import Estimator
+
+import pickle
+from tqdm import tqdm
+from datasets import load_dataset
+from collections import defaultdict
+import random
+
+from transformers import AutoTokenizer
+
+import spacy
+nlp = spacy.load('en_core_web_sm')
+
+
+#calculating tokens idf on 1M documents from RedPajama dataset 
+def calcu_idf(tokenizer_path, path):
+    print(tokenizer_path)
+    filename = "/tmp/not_exist/filenames.pkl"
+    os.makedirs(os.path.dirname(path), exist_ok=True)
+    dataset = load_dataset("togethercomputer/RedPajama-Data-1T-Sample")
+    data = [d for d in dataset["train"]]
+    random.seed(42)
+    random.shuffle(data)
+    tokenizer = AutoTokenizer.from_pretrained(tokenizer_path, use_fast=True)
+    document_frequency = defaultdict(int)
+    offset = 1 if 'facebook' in tokenizer_path else 0
+    for doc in tqdm(data):
+        tokenized_doc = tokenizer(doc["text"])["input_ids"][offset:]
+        unique_tokens = set(tokenized_doc)
+        for token in unique_tokens:
+            document_frequency[token] += 1
+    total_documents = len(data)
+    print(np.array([math.log(total_documents / (document_frequency[i] + 1)) for i in range(len(tokenizer.vocab))]))
+    pickle.dump(np.array([math.log(total_documents / (document_frequency[i] + 1)) for i in range(len(tokenizer.vocab))]), open(path, "wb"))
+
+
+class Focus(Estimator):
+    def __init__(
+        self,
+        gamma: float = 0.9,
+        p: float = 0.01,
+        model_name: str = "meta-llama/Llama-3.2-1B",
+    ):
+        super().__init__([
+                            "greedy_log_probs",
+                            "greedy_tokens", 
+                            "greedy_texts",
+                            "attention_all",
+                            "tokenizer",
+                         ], 
+                         "sequence")
+        #path where idf scores are stored 
+        path = f"../../../focus_data/token_idf_{model_name.split('/')[-1]}.pkl"
```

Comment:
Add this path to the parameters in the `__init__` function and specify in config